### PR TITLE
Feature/native maxmind

### PR DESCRIPTION
**Pull request by gedl**

### CODE REVIEW

File: lib/src/main/java/technology/dice/dicewhere/building/maxmindnative/MaxmindNativeDatabase.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
+package technology.dice.dicewhere.building.maxmindnative;
```
Comment (stykiaz) [inline lib/src/main/java/technology/dice/dicewhere/building/maxmindnative/MaxmindNativeDatabase.java]: open source header comment

File: lib/src/main/java/technology/dice/dicewhere/building/IPDatabase.java [hunk 2]
```diff
@@ -1,69 +1,9 @@
...
-              } catch (InvalidProtocolBufferException e) {
-                throw new RuntimeException(e);
-              }
-            });
-  }
 
-  public long size() {
-    return size;
-  }
+public interface IPDatabase {
```
Comment (stykiaz) [inline lib/src/main/java/technology/dice/dicewhere/building/IPDatabase.java]: copyright comment

File: lib/src/main/java/technology/dice/dicewhere/Main.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
+package technology.dice.dicewhere;
```
Comment (stykiaz) [inline lib/src/main/java/technology/dice/dicewhere/Main.java]: copyright comment

File: lib/src/main/java/technology/dice/dicewhere/Main.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+            .withProvider(
+                new MaxmindNativeSource(
+                    Paths.get("/Users/gluiz/Downloads/GeoIP2-City_20200121/GeoIP2-City.mmdb"),
+                    Paths.get(
+                        "/Users/gluiz/Downloads/GeoIP2-Anonymous-IP_20200124/GeoIP2-Anonymous-IP.mmdb")))
+            .build();
+    final Map<ProviderKey, Optional<IpInformation>> resolve =
+        build.resolve(new IP(InetAddress.getByName("82.102.20.51")));
+
+    int a = 2;
```
Comment (AndreyLebedenko) [inline lib/src/main/java/technology/dice/dicewhere/Main.java]: Why? Is it used?

File: lib/src/main/java/technology/dice/dicewhere/building/navigablemap/NavigableMapIpDatabase.java [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+              } catch (InvalidProtocolBufferException e) {
+                throw new RuntimeException(e);
+              }
+            });
+  }
+
+  public long size() {
+    return size;
+  }
```
Comment (AndreyLebedenko) [inline lib/src/main/java/technology/dice/dicewhere/building/navigablemap/NavigableMapIpDatabase.java]: Used only in tests. No need to create an extra field for tests only.

File: lib/src/main/java/technology/dice/dicewhere/reading/maxmind/MaxmindNativeSource.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+import technology.dice.dicewhere.building.DatabaseBuilderListener;
+import technology.dice.dicewhere.building.IPDatabase;
+import technology.dice.dicewhere.building.maxmindnative.MaxmindNativeDatabase;
+import technology.dice.dicewhere.lineprocessing.LineProcessorListener;
+import technology.dice.dicewhere.provider.ProviderKey;
+import technology.dice.dicewhere.provider.maxmind.MaxmindProviderKey;
+import technology.dice.dicewhere.reading.LineReaderListener;
+import technology.dice.dicewhere.reading.SourceReader;
+
+public class MaxmindNativeSource implements SourceReader {
```
Comment (AndreyLebedenko) [inline lib/src/main/java/technology/dice/dicewhere/reading/maxmind/MaxmindNativeSource.java]: Not unit tested.

File: lib/src/main/java/technology/dice/dicewhere/provider/maxmind/decorator/MaxmindVpnDecoratorDbReader.java [hunk 7]
```diff
@@ -33,7 +33,8 @@
 
   public MaxmindVpnDecoratorDbReader(
```
Comment (AndreyLebedenko) [inline lib/src/main/java/technology/dice/dicewhere/provider/maxmind/decorator/MaxmindVpnDecoratorDbReader.java]: DecoratorDbReader and it's descendants are not thread safe, due to non-volatile, non-atomic `lastFetched`

**Comment by AndreyLebedenko** [comment]
Not sure if it is critical or not, but the project can not be built on Windows, due to use of `.getFile()` in unit tests. For example `IPResolverTest`. If build on Windows is required, those must be replaced with `.toURI()`.

File: lib/src/main/java/technology/dice/dicewhere/Main.java (hunk 4 above)
Comment (gedl) [reply lib/src/main/java/technology/dice/dicewhere/Main.java]: The whole file is spurious. removed.

File: lib/src/main/java/technology/dice/dicewhere/building/navigablemap/NavigableMapIpDatabase.java (hunk 5 above)
Comment (gedl) [reply lib/src/main/java/technology/dice/dicewhere/building/navigablemap/NavigableMapIpDatabase.java]: It's actually part of the intended API.

File: lib/src/main/java/technology/dice/dicewhere/provider/maxmind/decorator/MaxmindVpnDecoratorDbReader.java (hunk 7 above)
Comment (gedl) [reply lib/src/main/java/technology/dice/dicewhere/provider/maxmind/decorator/MaxmindVpnDecoratorDbReader.java]: Please open an issue to address this

**Comment by gedl** [comment]
>Not sure if it is critical or not, but the project can not be built on Windows, due to use of .getFile() >in unit tests. For example IPResolverTest. If build on Windows is required, those must be replaced >with .toURI().

Please open an issue to address this

**Pull request merged**